src/Error/SourceExceptionFactory.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Error;
+
+/**
+ * @internal
+ */
+final class SourceExceptionFactory
+{
+    /**
+     * @param array{class: class-string<\Throwable>, message: string, code: int, file: string, line: int} $error
+     */
+    public static function fromArray(array $error): \Throwable
+    {
+        $exceptionClass = $error['class'];
+
+        try {
+            $exception = new $exceptionClass($error['message'], $error['code']);
+
+            if (
+                $exception->getMessage() !== $error['message']
+                || $exception->getCode() !== $error['code']
+            ) {
+                throw new \RuntimeException('Failed to create exception from array. Message and code are not the same.');
+            }
+        } catch (\Throwable $e) {
+            $exception = new \RuntimeException(
+                sprintf('[%s] %s', $exceptionClass, $error['message']),
+                $error['code']
+            );
+        }
+
+        try {
+            $exceptionReflection = new \ReflectionClass($exception);
+            foreach (['file', 'line'] as $property) {
+                $propertyReflection = $exceptionReflection->getProperty($property);
+                $propertyReflection->setAccessible(true);
```

Comment:
yeah, i was trying to avoid that, same as you, by instantiating `new self()` and not `new RuntimeException`. 

non-blocking

Reply:
Yes, but you were using class that was a child of `RuntimeException` so you could do that, I use factory that is not exception itself. I believe my approach is better since it supports errors too, that are separate branch of throwables (I'll add test for `ParseError`).